hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -0,0 +1,301 @@
+
+            // TODO: Kick off special file handling if needed
+
+            // TODO: Finalize transaction with the help of the token service
+
+            // commit state
+            stack.commit();
+        } catch (PreCheckException e) {
+            recordFailedTransaction(e.responseCode(), recordBuilder, recordListBuilder);
+        } catch (HandleException e) {
+            recordFailedTransaction(e.getStatus(), recordBuilder, recordListBuilder);
+        } catch (Throwable e) {
+            LOG.error("An unexpected exception was thrown during handle", e);
+            recordBuilder.status(ResponseCodeEnum.UNKNOWN);
+        }
+
+        // TODO update receipt
+
+        // TODO: handle long scheduled transactions
+
+        // TODO: handle system tasks
+
+        // store all records at once
+        recordManager.endUserTransaction(recordListBuilder.build());
+    }
+
+    private void recordFailedTransaction(
+            @NonNull final ResponseCodeEnum status,
+            @NonNull final SingleTransactionRecordBuilder recordBuilder,
+            @NonNull final RecordListBuilder recordListBuilder) {
+        recordBuilder.status(status);
+        recordListBuilder.revertChildRecordBuilders(recordBuilder);
+        // TODO: Finalize failed transaction with the help of token-service and commit required state changes
+    }
+
+    /*
+     * This method gets all the verification data for the current transaction. If pre-handle was previously ran
+     * successfully, we only add the missing keys. If it did not run or an error occurred, we run it again.
+     * If there is a due diligence error, this method will return a CryptoTransfer to charge the node along with
+     * its verification data.
+     */
+    @NonNull
+    private VerificationResult getVerifications(
+            @NonNull final HederaState state,
+            @NonNull final ConsensusEvent platformEvent,
+            @NonNull final ConsensusTransaction platformTxn)
+            throws PreCheckException {
+        final var metadata = platformTxn.getMetadata();
+        // We do not know how long transactions are kept in memory. Clearing metadata to avoid keeping it for too long.
+        platformTxn.setMetadata(null);
+
+        // First check if pre-handle was run successfully and all configuration has not changed
+        if (preHandleStillValid(metadata)) {
+            final var previousResult = (PreHandleResult) metadata;
+            // In case of due diligence error, we prepare a CryptoTransfer to charge the node and return immediately.
+            if (previousResult.status() == Status.NODE_DUE_DILIGENCE_FAILURE) {
+                return createPenaltyPayment();
+            }
+
+            // If pre-handle was successful, we need to add signatures that were not known at the time of pre-handle.
+            if (previousResult.status() == Status.SO_FAR_SO_GOOD) {
+                return addMissingSignatures(state, previousResult);
+            }
```

Comment:
This is probably correct. However, I wonder whether it would be safer to use a switch and cover all cases so if another status is added in the future, we don't forget to deal with it here.

Reply:
Currently, we do not cover all cases. If `status()` returns `PRE_HANDLE_FAILURE` or `UNKNOWN_FAILURE`, we want to leave the if-branch and run the code at the end of the method. It should be rewritten to make this more obvious.